test/jdk/java/net/Socket/ConnectFailTest.java

```diff
@@ -0,0 +1,268 @@
+    }
+
+    /**
+     * Verifies socket is closed when {@code unboundSocket.connect()} fails.
+     * @param socket an unbound socket
+     */
+    @ParameterizedTest
+    @MethodSource("boundSockets")
+    void testBoundSocket(Socket socket) throws IOException {
+        try (socket) {
+            assertTrue(socket.isBound());
+            assertFalse(socket.isConnected());
+            assertThrows(IOException.class, () -> socket.connect(REFUSING_SOCKET_ADDRESS));
+            assertTrue(socket.isClosed());
+        }
+    }
+
+    @Test
+    void testConnectedSocket() throws Throwable {
+        testConnectedSocket(
+                ConnectFailTest::createConnectedSocket,
+                executable -> {
+                    SocketException exception = assertThrows(SocketException.class, executable);
+                    assertEquals("already connected", exception.getMessage());
+                });
+    }
+
+    @Test
+    void testConnectedNioSocket() throws Throwable {
+        testConnectedSocket(
+                ConnectFailTest::createConnectedNioSocket,
+                executable -> assertThrows(AlreadyConnectedException.class, executable));
+    }
+
+    /**
+     * Verifies socket is not closed when {@code `connectedSocket.connect()`} fails.
+     * @param connectedSocketFactory a connected socket factory
+     * @param reconnectFailureVerifier a consumer verifying the thrown reconnect failure
+     */
+    private static void testConnectedSocket(
+            Function<SocketAddress, Socket> connectedSocketFactory,
+            Consumer<Executable> reconnectFailureVerifier)
+            throws Throwable {
+        withEphemeralServerSocket(serverSocket -> {
+            SocketAddress serverSocketAddress = serverSocket.getLocalSocketAddress();
+            try (Socket socket = connectedSocketFactory.apply(serverSocketAddress)) {
+                assertTrue(socket.isBound());
+                assertTrue(socket.isConnected());
+                // `Socket` and `SocketChannel` differ in how they fail on re-connection attempts on an already
+                // connected socket. Hence, we delegate this particular check:
+                reconnectFailureVerifier.accept(() -> socket.connect(REFUSING_SOCKET_ADDRESS));
+                assertFalse(socket.isClosed());
+            }
+        });
+    }
+
+    /**
+     * Verifies socket is closed when {@code unboundSocket.connect(unresolvedAddress)} fails.
+     * @param socket an unbound socket
+     */
+    @ParameterizedTest
+    @MethodSource("unboundSockets")
+    void testUnboundSocketWithUnresolvedAddress(Socket socket) throws IOException {
+        try (socket) {
+            assertFalse(socket.isBound());
+            assertFalse(socket.isConnected());
+            assertThrows(IOException.class, () -> socket.connect(UNRESOLVED_ADDRESS));
+            assertTrue(socket.isClosed());
+        }
+    }
+
+    @SuppressWarnings("resource")
```

Comment:
What is this SuppressWarnings for?

Reply:
For IntelliJ's `'SocketChannel' used without 'try'-with-resources statement` warning on `open()` method. Am I supposed to not suppress IDE warnings?

Reply:
Oh - OK - I see. That's fine. We usually don't suppress warnings in tests. Maybe we should.

Reply:
:+1: Switched to a more fine-grained suppression.